spike_main/spike.cc

```diff
@@ -93,6 +93,44 @@ static void read_file_bytes(const char *filename,size_t fileoff,
   in.read(read_buf, read_sz);
 }
 
+bool sort_mem_region(const std::pair<reg_t, mem_t*> &a,
+                       const std::pair<reg_t, mem_t*> &b)
+{
+  if (a.first == b.first)
+    return (a.second->size() < b.second->size());
+  else
+    return (a.first < b.first);
+}
+
+void check_memory_regions(std::vector<std::pair<reg_t, mem_t*>>& mems)
+{
+  // check the user specified memory regions and merge the overlapping or
+  // eliminate the containing parts
+  std::sort(mems.begin(), mems.end(), sort_mem_region);
+  int32_t start_page = 0, end_page = 0;
+  std::vector<std::pair<reg_t, mem_t*>>::reverse_iterator it = mems.rbegin();
+  std::vector<std::pair<reg_t, mem_t*>>::reverse_iterator _it = mems.rbegin();
+  for(; it != mems.rend(); ++it) {
+    int32_t _start_page = it->first/PGSIZE;
+    int32_t _end_page = _start_page + it->second->size()/PGSIZE;
```

Comment:
This will fail for physical addresses greater than 44 bits.  Should use reg_t, not int32_t.